### PR TITLE
Initial support for buildEnvs in manifest.yaml

### LINKED ISSUE #640: Support for BuildEnv in manifest.yaml to be propagated to func.yaml
Currently, BuildEnv is not supported inside `manifest.yaml` in templates and some cases require setting some Build environment variables for buildpacks to work. This issue was discussed here: https://github.com/knative-sandbox/kn-plugin-func/pull/631#pullrequestreview-797818776

### PR DESCRIPTION
**Pull request by salaboy**
# Changes

:gift: support for BuildEnvs inside `manifest.yaml`

/kind enhancement

Fixes #640

### CODE REVIEW

**Comment by salaboy** [comment]
@lance I know that you might be super busy, but feedback is appreciated.

File: templates_test.go [hunk 1]
```diff
@@ -377,3 +380,55 @@ func TestTemplateModeRemote(t *testing.T) {
 }
 
 // TODO: test typed errors for custom and remote (embedded checked)
+
+func TestTemplateFuncYAML(t *testing.T) {
```
Comment (lkingland) [inline templates_test.go:384]: It helps quite a bit to have all tests include a statement of purpose as their doc string, and a function name that is as similar as possible. 
Perhaps something like the below?
```suggestion
// TestTemplateManifestBuildEnvs ensures that BuildEnvs specified in a
// template's manifest are included in the final Function.
func TestTemplateManifestBuildEnvs(t *testing.T) {a
```

**Review by lance — COMMENTED** [review]
/lgtm

The PR would benefit from tests at runtime and repo level.

File: templates_test.go [hunk 2]
```diff
@@ -377,3 +380,55 @@ func TestTemplateModeRemote(t *testing.T) {
...
+	}
+
+	config := fn.Config{}
+	err = yaml.Unmarshal(funcByteArray, &config)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	testVariableName := "TEST_VARIABLE"
+	testVariableValue := "test"
```
Comment (lkingland) [inline templates_test.go:422]: This is good, but this is really testing a side-effect.  The intent is to ensure that BuildEnvs set in the Template are affected on the resultant Function, so you could actually save a little work, and test more of the process overall, by perhaps creating a Function from the Template and validating it looks as expected:
```go
f, err := fn.NewFunction(root)
if err != nil {
  t.Fatal(err)
}
if !reflect.DeepEqual(f.BuildEnvs, envs) {
```
.. and then testing if f.BuildEnvs is

**Review by lkingland — COMMENTED** [review]
Looks great!
Just one minor suggested way to expand the test to make it validate the whole cycle (template manifest -> final Funciton).
Also thanks for including the spelling-corrections!

**Comment by salaboy** [comment]
@lance @lkingland thanks a lot for the feedback... I will be working on your comments today

File: templates_test.go (hunk 1 above)
Comment (salaboy) [reply templates_test.go:384]: 👍

**Comment by salaboy** [comment]
@lance @lkingland all changes added.. thanks a lot for the suggestions.. now the test checks variables at repo, runtime and template level using the `go-cmp` to check the different `buildEnvs` depending the project. This should enable the SpringNative templates to work in #630

**Review by zroubalik — COMMENTED** [review]
/lgtm

**Comment by lkingland** [comment]
> @lance @lkingland all changes added.. thanks a lot for the suggestions.. now the test checks variables at repo, runtime and template level using the `go-cmp` to check the different `buildEnvs` depending the project. This should enable the SpringNative templates to work in #630

LGTM!  Except that the unit test says that something is on fire.  You should be able to run `make test` locally and recreate the issue.  Please comment if there's any problems, we're in the midst of a transition of build tooling!

**Comment by salaboy** [comment]
@lance the unit tests are working just fine in my local environment, can you trigger a re-test?

```
salaboy> make test
go test -race -cover -coverprofile=coverage.out ./...
ok  	knative.dev/kn-plugin-func	16.895s	coverage: 71.3% of statements
?   	knative.dev/kn-plugin-func/buildpacks	[no test files]
ok  	knative.dev/kn-plugin-func/cloudevents	6.885s	coverage: 70.6% of statements
ok  	knative.dev/kn-plugin-func/cmd	2.861s	coverage: 23.1% of statements
?   	knative.dev/kn-plugin-func/cmd/func	[no test files]
ok  	knative.dev/kn-plugin-func/docker	11.944s	coverage: 43.2% of statements
?   	knative.dev/kn-plugin-func/k8s	[no test files]
ok  	knative.dev/kn-plugin-func/knative	2.428s	coverage: 3.8% of statements
?   	knative.dev/kn-plugin-func/mock	[no test files]
?   	knative.dev/kn-plugin-func/plugin	[no test files]
?   	knative.dev/kn-plugin-func/progress	[no test files]
?   	knative.dev/kn-plugin-func/schema/generator	[no test files]
?   	knative.dev/kn-plugin-func/ssh	[no test files]
ok  	knative.dev/kn-plugin-func/utils	0.280s	coverage: 100.0% of statements
```

**Comment by salaboy** [comment]
@lance @lkingland @zroubalik apologies for the broken test. It is now working! :)

**Comment by salaboy** [comment]
@lance @lkingland 
This is actually working in my local env: 
```
salaboy> make test
# Generating pkged.go using pkger
./hack/update-pkger.sh
go test -race -cover -coverprofile=coverage.out ./...
ok  	knative.dev/kn-plugin-func	11.624s	coverage: 71.8% of statements
?   	knative.dev/kn-plugin-func/buildpacks	[no test files]
ok  	knative.dev/kn-plugin-func/cloudevents	6.747s	coverage: 70.6% of statements
ok  	knative.dev/kn-plugin-func/cmd	3.757s	coverage: 23.1% of statements
?   	knative.dev/kn-plugin-func/cmd/func	[no test files]
ok  	knative.dev/kn-plugin-func/docker	8.647s	coverage: 43.2% of statements
?   	knative.dev/kn-plugin-func/k8s	[no test files]
ok  	knative.dev/kn-plugin-func/knative	2.811s	coverage: 3.8% of statements
?   	knative.dev/kn-plugin-func/mock	[no test files]
?   	knative.dev/kn-plugin-func/plugin	[no test files]
?   	knative.dev/kn-plugin-func/progress	[no test files]
?   	knative.dev/kn-plugin-func/schema/generator	[no test files]
?   	knative.dev/kn-plugin-func/ssh	[no test files]
ok  	knative.dev/kn-plugin-func/utils	0.191s	coverage: 100.0% of statements
```

**Comment by lkingland** [comment]
@salaboy I am seeing this in my environment as well.

In PR #655 I removed the plural types `Envs` `Labels` etc. in favor of their direct slice counterparts.  For some reason my local build continued to compile the `cmd` package, despite these types being gone from the codebase.  It tried manually resetting my local build and test caches:
```
go clean -modcache -testcache -cache
```
.. yet it continued to compile despite the missing types!

Perhaps there is a need to manually update the module via `go get -d knative.dev/kn-plugin-func`?  Please let me know if that fixes the problem, as I would like to know.

For now, you should be able to simply replace the type `Envs` with `[]Env`, and it will succeed both locally and in CI.

**Comment by salaboy** [comment]
@lkingland I couldn't reproduce the issue in my environment.. but I've just made the changes from `Envs` to `[]Env` let's see if that works.

**Comment by salaboy** [comment]
@lkingland it looks like the Env issue is fixed, but did someone remove the `using(t, root)()` function?

**Comment by salaboy** [comment]
ok rebasing should work now :) `Using` it is

**Pull request merged**